File: demo/SwgTatooine/World/Rng.cs
using System;

namespace SwgTatooine;

/// <summary>
/// A small deterministic generator, passed by <c>ref</c> so a whole world build draws from one stream and is reproducible
/// from the seed alone.
/// </summary>
/// <remarks>
/// xorshift32 rather than <see cref="Random"/> because the world build draws millions of numbers at high population
/// multipliers and this is a handful of instructions with no allocation and no virtual call. Determinism is the point,
/// not statistical quality: two arms of a partitioning sweep must see the identical world or the comparison says nothing.
/// </remarks>
public struct Rng
{
    private uint _state;

    public Rng(uint seed) => _state = seed == 0 ? 0x9E3779B9u : seed;

    /// <summary>Next raw 32-bit value.</summary>
    public uint Next()
    {
        _state ^= _state << 13;
        _state ^= _state >> 17;
        _state ^= _state << 5;
        return _state;
    }

    /// <summary>Uniform in [0, 1).</summary>
    public float NextFloat() => (Next() >> 8) * (1f / 16_777_216f);

    /// <summary>Uniform in [min, max).</summary>
    public float NextRange(float min, float max) => min + ((max - min) * NextFloat());

    /// <summary>Uniform integer in [min, max).</summary>
    public int NextInt(int min, int max) => max <= min ? min : min + (int)(Next() % (uint)(max - min));

    /// <summary>
    /// A point uniform BY AREA inside a disc. The square root is what makes it uniform — sampling the radius linearly
    /// would pile two thirds of a city into its middle third and give every city a false density gradient.
    /// </summary>
    public (float X, float Z) PointInDisc(float cx, float cz, float radius)
    {
        var a = NextFloat() * MathF.PI * 2f;
        var r = radius * MathF.Sqrt(NextFloat());
        return (cx + (MathF.Cos(a) * r), cz + (MathF.Sin(a) * r));
    }
}

File: demo/SwgTatooine.Client/src/data/mock/rng.ts
/** Deterministic PRNG (mulberry32): a seed reproduces the whole mock world and its behaviour. */
export class Rng {
  private state: number;

  constructor(seed: number) {
    this.state = seed >>> 0 || 0x9e3779b9;
  }

  /** Uniform in [0, 1). */
  next(): number {
    this.state = (this.state + 0x6d2b79f5) >>> 0;
    let t = this.state;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  range(min: number, max: number): number {
    return min + (max - min) * this.next();
  }

  /** Integer in [min, max). */
  int(min: number, max: number): number {
    return min + Math.floor((max - min) * this.next());
  }

  /** A point uniform by area in a disc; written to `out[0]`, `out[1]`. */
  pointInDisc(cx: number, cz: number, radius: number, out: Float64Array): void {
    const a = this.next() * Math.PI * 2;
    const r = radius * Math.sqrt(this.next());
    out[0] = cx + Math.cos(a) * r;
    out[1] = cz + Math.sin(a) * r;
  }
}
